launch/launch/actions/execute_local.py

```diff
@@ -420,27 +438,21 @@ def __on_shutdown(self, event: Event, context: LaunchContext) -> Optional[SomeAc
             send_sigint=not due_to_sigint or context.noninteractive,
         )
 
-    def __get_shutdown_timer_actions(self) -> List[Action]:
+    def __get_shutdown_timer_actions(self, context) -> List[Action]:
         base_msg = \
             "process[{}] failed to terminate '{}' seconds after receiving '{}', escalating to '{}'"
 
-        def printer(context, msg, timeout_substitutions):
-            self.__logger.error(msg.format(
-                context.locals.process_name,
-                perform_substitutions(context, timeout_substitutions),
-            ))
+        def printer(context, msg):
+            self.__logger.error(msg.format(context.locals.process_name))
 
-        sigterm_timeout = self.__sigterm_timeout
-        sigkill_timeout = [PythonExpression(
-            ('float(', *self.__sigterm_timeout, ') + float(', *self.__sigkill_timeout, ')')
-        )]
         # Setup a timer to send us a SIGTERM if we don't shutdown quickly.
+        sigterm_timeout = self.__sigterm_timeout_value
         self.__sigterm_timer = TimerAction(
             period=sigterm_timeout,
             actions=[
                 OpaqueFunction(
                     function=printer,
-                    args=(base_msg.format('{}', '{}', 'SIGINT', 'SIGTERM'), sigterm_timeout)
+                    args=(base_msg.format('{}', sigterm_timeout, 'SIGINT', 'SIGTERM'), )
```

Comment:
These changes are concerning to me because if `__get_shutdown_timer_actions` is ever called when not actually executing the actions then it seems like it might fail. Like can you ensure that `self.__sigterm_timeout` has been assigned a value at this point?

Reply:
You might be right that these didn't need to be deferred as substitutions, but I'm not 100% certain of that. Can you expand on it?

Reply:
`__get_shutdown_timer_actions()` is getting `context` as an argument, so in any case the substitutions could be resolved here.
This PR is instead resolving the substitutions in `execute()`:

https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L728-L729

I think that's fine, because `__get_shutdown_timer_actions()` is only called from [_shutdown_process()](https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L253), which is called from two places [__on_shutdown_process_event()](https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L290) and [__on_shutdown](https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L434).
Those are used [here](https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L700) and [here](https://github.com/ros2/launch/blob/e326c0dfee10ad1beaa7c12db6d97eb15da57b74/launch/launch/actions/execute_local.py#L715), which are event handlers defined in `execute()`.
So I think that `__sigterm_timeout_value` and `__sigkill_timeout_value` are guaranteed to have a value here.

I also think that `execute()` is the place were these substitutions should be resolved.
If they were resolved in an event handler afterwards, the value of the launch configuration could change, and that could lead to really confusing results.

Reply:
Since the `__get_shutdown_timer_actions()` function now takes a `context` object, I think this change is safe because callers will need to source the context object which is only ever available from the `execute` function, but I don't think it was necessary for this work.

The point about the event handler and the right time to evaluate the substitutions is fine, but it could have been done without resolving the substitutions in this function, keeping it declarative (not requiring a context in its scope).